### PR TITLE
(revised) initial cmake-based build setup for Omega

### PR DESCRIPTION
**Pull request by grnydawn**
This PR includes the initial cmake-based build-setup for Omega.

This PR replaces `https://github.com/E3SM-Project/Omega/pull/17` based on the comments and suggestions of the PR.

Omega build supports two modes:
* Standalone
* E3SM component

The main purpose of this PR is to initiate discussions on Omega build design and implementation.

Checklist
* [x] Documentation:
  * [ ] Design document has been generated and added to the docs
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/omega/develop/developers_guide/building_docs.html) and changes look as expected
* [x] Testing
  * [ ] A comment in the PR documents testing used to verify the changes including any tests that are added/modified/impacted.
  * [x] CTest unit tests for new features have been added per the approved design. 
  * [ ] Polaris tests for new features have been added per the approved design (and included in a test suite)
  * [ ] Unit tests have passed. Please provide a relevant CDash build entry for verification.
  * [ ] Polaris test suite has passed
  * [ ] Performance related PRs: Please include a relevant PACE experiment link documenting performance before and after.
* [ ] Stealth Features
  * [ ] If any stealth features are included in the PR, please confirm that they have been documented.

### CODE REVIEW

File: components/omega/OmegaBuild.cmake
Comment (sarats) [inline components/omega/OmegaBuild.cmake:7]: Checking: Do we want CamelCase for library name or just lowercase (like in exe)?

Comment (philipwjones) [reply components/omega/OmegaBuild.cmake:7]: Yes, we're following llvm standards that use CamelCase with initial capital letter for classes and that follows to most filenames and it makes sense to propagate up to library name too.

File: components/omega/doc/userGuide/OmegaBuild.md
Comment (philipwjones) [inline components/omega/doc/userGuide/OmegaBuild.md:10]: Can we add a sentence or two with the need to create a build directory. Maybe something like:

CMake and OMEGA prefer an out-of-source build.  This enables a user to build and maintain multiple executables from the same source directory. The standard practice is for a user to create a separate directory where the build should take place and the commands below should be launched from that directory.

**Comment by philipwjones** [comment]
@grnydawn  I've tried to build using this on both Chrysalis and Crusher and it seems like the PATH is not getting propagated to the YAKL portion of the build. So it starts out finding the compiler just fine:
-- Check for working CXX compiler: /gpfs/fs1/soft/chrysalis/spack/opt/spack/linux-centos8-x86_64/gcc-9.3.0/intel-20.0.4-kodw73g/bin/icpc - skipped

But when it gets to the YAKL part of the build, it fails with:
CMake Error at /home/ac.pjones/OMEGAcmake/externals/YAKL/CMakeLists.txt:1 (project):
  The CMAKE_CXX_COMPILER:

    icpc

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.

This also happens with an MPI-wrapper version (mpicxx). It recognizes the CMAKE_CXX_COMPILER name just fine, just can't seem to locate it, even though it's in the PATH.

File: components/omega/OmegaBuild.cmake
Comment (grnydawn) [reply components/omega/OmegaBuild.cmake:7]: I agree to use CamelCase for lib names. The executable name does not use CamelCase because I thought it is convention to use lowercase for exe name, but if preferred to use CamelCase, I am good to change.

File: components/omega/doc/userGuide/OmegaBuild.md
Comment (grnydawn) [reply components/omega/doc/userGuide/OmegaBuild.md:10]: Sure, I will add the info in next commit.

**Comment by grnydawn** [comment]
> Also note that if we don't use the Fortran interfaces for YAKL, we don't need to build and actually only need to add the YAKL source and YAKL/external directories so that we can pick up the header files. So if the YAKL build is problematic, we can simply add the YAKL dirs to the include paths and it should work.

I agree that we may only need to add the directories for YAKL header files because YAKL moved to header-only library as far as I know.

**Comment by grnydawn** [comment]
@philipwjones I fixed the compiler detection issue by running `which` linux command to get the full path of the compiler specified in OMEGA_CXX_COMPILER. It worked with `CC` and `mpic++` on Crusher for me.

The fix also correctly built YAKL library. Eventually, I agree to use only YAKL source files and headers. But, I think I will leave to build YAKL for now until I found proper way to exclude the YAKL library build.

**Comment by grnydawn** [comment]
@xylar, Could you suggest me how to fix the CI/CD check? It looks that the failure happens at running clang-format.

**Comment by xylar** [comment]
@grnydawn, can you rebase and pick up the latest CI changes made in #22?  Hopefully, that will help.

File: components/omega/doc/userGuide/OmegaBuild.md
Comment (philipwjones) [inline components/omega/doc/userGuide/OmegaBuild.md:76]: Ok, one last bit here for standalone builds. We should probably have text here to point to where the omega.exe ends up after a build. For both cases with and without OMEGA_INSTALL_PREFIX. And add the "make install" step for the latter.

**Comment by xylar** [comment]
@grnydawn, the fact that the commits from `main` show up here sugests you rebased in the wrong direction or cherry-picked the commits or something like that.  I don't think that's a good idea because the same commits will show up in the commit history in 2 places.

But it doesn't seem to have fixed the CI errors and I also don't understand why not.

**Comment by xylar** [comment]
I will try to figure out how to fix CI in #24.  @grnydawn, you can also look there to see what I had in mind with rebasing onto `develop`

File: components/omega/doc/userGuide/OmegaBuild.md
Comment (grnydawn) [reply components/omega/doc/userGuide/OmegaBuild.md:95]: I updated the user guide with the explanation of the output paths and the usage of OMEGA_INSTALL_PREFIX.

**Comment by grnydawn** [comment]
> 

Ok. Thanks for letting me know. I will pay more attention on that.

**Comment by grnydawn** [comment]
@xylar @philipwjones , Xylar's update in #25 did resolve the CI checks in this PR.

But, because I already messed this branch up with git rebase  and because this branch has unnecessary many small commits, I think I will create another PR with cleaner commit history Sorry for the duplications.

**Comment by xylar** [comment]
@grnydawn, I would recommend that you just force push to this branch. I don't think there is any advantage to making a new PR.

```
git push --force ...
```

**Comment by xylar** [comment]
That is, make exactly the changes in this branch that you have in mind for a new PR and instead force push to this branch.

**Comment by xylar** [comment]
@grnydawn rather than merging develop inti this branch, use

```
git rebase -i E3SM-Project/Omega/develop
```
The name of the E3SM-Project/Omega remote may be different for you.

This will lead to a cleaner result.

**Comment by grnydawn** [comment]
@xylar I had a merge conflict with .clang-format and tried to resolve the conflict by coping the recent version of the file. But after the merge, I got the same issue with CI check in this branch.

I guess that using the rebase command that you suggested is a right way to do, but I am a bit concerned that my commit tree on this branch is already different from E3SM-Project/Omega/develop and I would like to avoid any further concern from those differences. So, I think I would create a new branch based on the latest develop branch, and a new PR from the branch this time if that is allowed.

**Comment by xylar** [comment]
@grnydawn, you can create a new branch off of develop.  But there is no reason to make a new PR.  You can do whatever you want on the new branch.  Then, when you are happy with it, hard-reset this branch to match that branch and then force push it.  That way, this PR is identical to whatever you would do on the new PR. Opening new PRs just causes more confusion.
```
# after <other_branch> is in the state you want:
git reset --hard <other_branch> ykim/omeg/cmakebuild
git push --force  grnydawn/Omega ykim/omeg/cmakebuild
```

**Comment by xylar** [comment]
@grnydawn, if you prefer, we can chat more about this on Slack.

**Comment by grnydawn** [comment]
@xylar Ok. once I have my new branch ready, I will try to follow the git command that you suggested. Let' chat over Slack more on this. Thanks!

**Review by xylar — APPROVED** [review]
Looks good to me based on my limited understanding of CMake.  @grnydawn, thanks for linting so CI is happy.

I will enable other linting checks in a follow-up PR after this goes in.  Let's not merge this until #25 goes in to avoid inadvertently merging that one as part of this one.

**Review by amametjanov — APPROVED** [review]

**Review by philipwjones — COMMENTED** [review]
@grnydawn  I've been able to build successfully with the latest changes.

For the docs, can you add the entries to the top level index.md file (in the doc directory). For the user guide, add userGuide/OmegaBuild in between the entries for QuickStart and DataTypes.  For the developer guide, just add the entry devGuide/CMakeBuild after the BuildDocs entry. With those changes, the docs build and they look good. 

After that change and probably a final rebase to clean up, I think this is good. I'll approve once those mods are committed.

**Comment by xylar** [comment]
Thanks for catching those issues with the docs, @philipwjones.

**Comment by xylar** [comment]
@grnydawn, please do see if you can build the docs (either locally on your laptop/desktop or on HPC, in which case you will need to copy the `doc/_build/html` directory to your laptop/desktop) and make sure your additions look okay.  Otherwise, we end up having to do a follow-up PR, which is a bit of a hassle.

**Comment by grnydawn** [comment]
> @grnydawn I've been able to build successfully with the latest changes.
> 
> For the docs, can you add the entries to the top level index.md file (in the doc directory). For the user guide, add userGuide/OmegaBuild in between the entries for QuickStart and DataTypes. For the developer guide, just add the entry devGuide/CMakeBuild after the BuildDocs entry. With those changes, the docs build and they look good.
> 
> After that change and probably a final rebase to clean up, I think this is good. I'll approve once those mods are committed.

@philipwjones , I added the entries in "doc/index.md".

**Comment by grnydawn** [comment]
> @grnydawn, please do see if you can build the docs (either locally on your laptop/desktop or on HPC, in which case you will need to copy the `doc/_build/html` directory to your laptop/desktop) and make sure your additions look okay. Otherwise, we end up having to do a follow-up PR, which is a bit of a hassle.

@xylar Thanks for the comment. I had to change section levels to remove some sphinx complaints.

**Comment by xylar** [comment]
>  I had to change section levels to remove some sphinx complaints.

I think it's fine to leave those complaints for now.  We can fix those later when we have more comprehensive documentation that these smaller documents become part of.  But I appreciate you doing that.

**Review by philipwjones — APPROVED** [review]
Thanks for the changes @grnydawn  I'm good with this now

**Comment by grnydawn** [comment]
> Thanks for the changes @grnydawn I'm good with this now

@philipwjones great! I added you as an assignee because I cannot merge this branch into develop. So, if you are okay with it, please go ahead and merge.

**Pull request merged**
